File: bot.py
"""
bot.py
------
This is intentionally a thin "front door" now that the Mini App handles
almost everything (registering, browsing, editing your listing, admin
tools). The bot's only real jobs are:
  - Greet people and hand them a button into the Mini App
  - Give admins a few text-command fallbacks that don't require opening
    the app (handy for a quick /stats or /broadcast from your phone)

Everything else — /register, /find, /services, /rate, /myprofile, etc. —
now lives in the Mini App (server.py + webapp/), which has a much better
interface for all of it than a chat conversation ever could.

COMMANDS:
  /start   -> welcome message + button to open the Mini App
  /app     -> opens the Mini App directly
  /help    -> what this bot does, and where the real features live

SETUP:
  1. pip install -r requirements.txt
  2. Message @BotFather on Telegram, run /newbot, copy the token it gives you
  3. Put that token in a file called token.txt (or set BOT_TOKEN env var)
  4. Run: python bot.py   (or python server.py to run the Mini App too)
"""

import os
import asyncio
import logging
from functools import wraps

from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup, WebAppInfo
from telegram.ext import Application, CommandHandler, ContextTypes, MessageHandler, filters

import database as db

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)


# ---- Load bot token ----
# Simplest option for learning: put your token directly in a file named
# token.txt (one line, nothing else). We read it here so you never have
# to paste your real token into code you might accidentally share.
def load_token():
    token = os.environ.get("BOT_TOKEN")
    if token:
        return token
    if os.path.exists("token.txt"):
        with open("token.txt") as f:
            return f.read().strip()
    raise RuntimeError(
        "No bot token found. Create a file named token.txt with your token "
        "inside, or set the BOT_TOKEN environment variable."
    )


# ---- Load admin Telegram IDs ----
# Admins are identified by their personal Telegram user ID (a number),
# NOT their username. To find your own ID, message @userinfobot on Telegram.
# Set as an environment variable ADMIN_IDS="123456789,987654321"
# or a file admins.txt with one ID per line (or comma-separated).
def load_admin_ids():
    ids_str = os.environ.get("ADMIN_IDS", "")
    if not ids_str and os.path.exists("admins.txt"):
        with open("admins.txt") as f:
            ids_str = f.read()

    root_admin_ids = set()
    for part in ids_str.replace("\n", ",").split(","):
        part = part.strip()
        if part.isdigit():
            root_admin_ids.add(int(part))

    # Root admins (from Render's ADMIN_IDS) + admins added on the fly via
    # /addadmin, which are stored in the database instead — that's what
    # lets any existing admin add another one instantly, no redeploy needed.
    return root_admin_ids | db.get_admin_ids_from_db()


def is_root_admin(telegram_id: int):
    """Root admins can only be changed via Render's ADMIN_IDS — /removeadmin can't touch them."""
    ids_str = os.environ.get("ADMIN_IDS", "")
    return str(telegram_id) in [p.strip() for p in ids_str.replace("\n", ",").split(",")]


# The Mini App URL (your Render URL). Set this as an environment variable
# once you know your deployed URL, e.g. https://your-app.onrender.com
WEBAPP_URL = os.environ.get("WEBAPP_URL", "")


def admin_only(handler_func):
    """
    A decorator that wraps a command handler so it only runs for admins.
    Put @admin_only directly above any command function you want to restrict.
    Non-admins get a polite refusal instead of the command running.
    """
    @wraps(handler_func)
    async def wrapper(update: Update, context: ContextTypes.DEFAULT_TYPE):
        if update.effective_user.id not in load_admin_ids():
            await update.message.reply_text("🔒 This command is for bot admins only.")
            return
        return await handler_func(update, context)
    return wrapper


def app_button_keyboard():
    """The 'Open App' button used on /start and /app."""
    return InlineKeyboardMarkup([
        [InlineKeyboardButton("🚀 Open App", web_app=WebAppInfo(url=WEBAPP_URL))]
    ])


# ---------- Basic commands ----------

async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if WEBAPP_URL:
        await update.message.reply_text(
            "👋 Welcome to VentureVault!\n\n"
            "Find trusted entrepreneurs and freelancers for any job, or list "
            "your own services and get discovered.\n\n"
            "Tap below to get started:",
            reply_markup=app_button_keyboard(),
        )
    else:
        await update.message.reply_text(
            "👋 Welcome! The app isn't set up yet — ask the bot owner to set the WEBAPP_URL environment variable."
        )


async def open_app(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not WEBAPP_URL:
        await update.message.reply_text(
            "The app isn't set up yet — the bot owner needs to set the WEBAPP_URL environment variable."
        )
        return
    await update.message.reply_text("Tap below to open the app:", reply_markup=app_button_keyboard())


async def help_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
    text = (
        "🧭 Almost everything lives in the app now — registering, browsing "
        "entrepreneurs, editing your listing, and ratings.\n\n"
        "/app — open it\n"
        "/start — welcome message"
    )
    if update.effective_user.id in load_admin_ids():
        text += "\n\nYou're an admin. Send /adminhelp to see admin-only commands (also available in the app's Admin tab)."

    await update.message.reply_text(text)


# ---------- Admin-only commands ----------
# Kept as a text fallback even though the Mini App's Admin tab covers the
# same ground — sometimes it's faster to fire off a quick /stats than to
# open the app.

@admin_only
async def admin_help(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(
        "🔧 *Admin commands*\n\n"
        "/stats — see totals (entrepreneurs, services, ratings)\n"
        "/broadcast <message> — message every registered entrepreneur\n"
        "/forceremove <name> — remove any entrepreneur's listing by name\n"
        "/addadmin <telegram_id> — make someone an admin instantly (no redeploy)\n"
        "/removeadmin <telegram_id> — remove a bot-added admin\n"
        "/listadmins — see everyone with admin access\n"
        "/verify <name> — mark a listing as manually verified (shows a badge)\n"
        "/unverify <name> — remove that badge",
        parse_mode="Markdown",
    )


@admin_only
async def stats(update: Update, context: ContextTypes.DEFAULT_TYPE):
    s = db.get_stats()
    await update.message.reply_text(
        f"📊 *Bot stats*\n"
        f"Entrepreneurs: {s['entrepreneurs']}\n"
        f"Unique services: {s['services']}\n"
        f"Ratings given: {s['ratings']}",
        parse_mode="Markdown",
    )


@admin_only
async def broadcast(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not context.args:
        await update.message.reply_text("Usage: /broadcast <message>")
        return

    message = " ".join(context.args)
    telegram_ids = db.get_all_telegram_ids()

    sent, failed = 0, 0
    for telegram_id in telegram_ids:
        try:
            await context.bot.send_message(chat_id=telegram_id, text=f"📢 Announcement:\n\n{message}")
            sent += 1
        except Exception as e:
            # Common reasons: the user blocked the bot, or never started a chat with it.
            logger.warning(f"Broadcast failed for {telegram_id}: {e}")
            failed += 1

    await update.message.reply_text(f"Broadcast complete. Sent: {sent}, failed: {failed}.")


@admin_only
async def force_remove(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not context.args:
        await update.message.reply_text("Usage: /forceremove <name>\nExample: /forceremove Jane Doe")
        return

    name = " ".join(context.args)
    success, telegram_id = db.force_delete_by_name(name)
    if success:
        await update.message.reply_text(f'✅ Removed the listing matching "{name}" (user ID {telegram_id}).')
    else:
        await update.message.reply_text(f'No entrepreneur found matching "{name}".')


@admin_only
async def add_admin_cmd(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not context.args or not context.args[0].isdigit():
        await update.message.reply_text(
            "Usage: /addadmin <telegram_id>\n"
            "The person you're adding must message @userinfobot to get their numeric ID first."
        )
        return

    new_admin_id = int(context.args[0])
    db.add_admin(new_admin_id, added_by=update.effective_user.id)
    await update.message.reply_text(f"✅ {new_admin_id} is now an admin — no redeploy needed.")


@admin_only
async def remove_admin_cmd(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not context.args or not context.args[0].isdigit():
        await update.message.reply_text("Usage: /removeadmin <telegram_id>")
        return

    target_id = int(context.args[0])
    if is_root_admin(target_id):
        await update.message.reply_text(
            "That's a root admin (set via Render's ADMIN_IDS) — remove them there instead, not here."
        )
        return

    removed = db.remove_admin(target_id)
    await update.message.reply_text("✅ Removed." if removed else "That ID wasn't a bot-added admin.")


@admin_only
async def list_admins_cmd(update: Update, context: ContextTypes.DEFAULT_TYPE):
    ids_str = os.environ.get("ADMIN_IDS", "")
    root_ids = {p.strip() for p in ids_str.replace("\n", ",").split(",") if p.strip()}
    db_ids = db.get_admin_ids_from_db()

    lines = ["👑 *Root admins* (Render ADMIN_IDS):"]
    lines += [f"• {i}" for i in root_ids] or ["  (none set)"]
    lines.append("\n🛠 *Added via /addadmin:*")
    lines += [f"• {i}" for i in db_ids] or ["  (none)"]
    await update.message.reply_text("\n".join(lines), parse_mode="Markdown")


async def handle_shared_contact(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """
    Fires when someone taps 'Verify with Telegram' in the Mini App, which
    triggers Telegram's native requestContact() prompt. If they approve,
    Telegram delivers their REAL phone number here as a message — this is
    what makes it trustworthy: Telegram itself vouches for the number,
    not just whatever the person typed into a form.
    """
    contact = update.message.contact
    # Only accept a contact if it's the person's OWN number, not one they
    # forwarded/shared on behalf of someone else in their contacts list.
    if contact.user_id != update.effective_user.id:
        await update.message.reply_text("Please share your own contact, not someone else's.")
        return

    updated = db.set_verified_phone(update.effective_user.id, contact.phone_number)
    if updated:
        await update.message.reply_text(
            "✅ Phone verified! Reopen the app (/app) to see the checkmark on your listing."
        )
    else:
        await update.message.reply_text(
            "You'll need to register first (via the app) before verifying your phone."
        )


@admin_only
async def verify_cmd(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not context.args:
        await update.message.reply_text("Usage: /verify <name>\nExample: /verify Jane Doe")
        return
    name = " ".join(context.args)
    success, telegram_id = db.set_identity_verified_by_name(name, True)
    await update.message.reply_text(
        f'✅ Marked "{name}" as verified.' if success else f'No entrepreneur found matching "{name}".'
    )


@admin_only
async def unverify_cmd(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not context.args:
        await update.message.reply_text("Usage: /unverify <name>")
        return
    name = " ".join(context.args)
    success, telegram_id = db.set_identity_verified_by_name(name, False)
    await update.message.reply_text(
        f'Removed verification from "{name}".' if success else f'No entrepreneur found matching "{name}".'
    )


def build_application():
    """
    Builds and configures the bot's Application object (registers every
    command handler) WITHOUT starting it. Kept separate from main() so
    server.py can reuse this exact setup when running the bot alongside
    the Mini App's web server.
    """
    db.init_db()
    token = load_token()

    app = (
        Application.builder()
        .token(token)
        .connect_timeout(30)
        .read_timeout(30)
        .get_updates_read_timeout(30)
        .build()
    )

    app.add_handler(CommandHandler("start", start))
    app.add_handler(CommandHandler("help", help_command))
    app.add_handler(CommandHandler("app", open_app))

    # Admin-only
    app.add_handler(CommandHandler("adminhelp", admin_help))
    app.add_handler(CommandHandler("stats", stats))
    app.add_handler(CommandHandler("broadcast", broadcast))
    app.add_handler(CommandHandler("forceremove", force_remove))
    app.add_handler(CommandHandler("addadmin", add_admin_cmd))
    app.add_handler(CommandHandler("removeadmin", remove_admin_cmd))
    app.add_handler(CommandHandler("listadmins", list_admins_cmd))
    app.add_handler(CommandHandler("verify", verify_cmd))
    app.add_handler(CommandHandler("unverify", unverify_cmd))
    app.add_handler(MessageHandler(filters.CONTACT, handle_shared_contact))

    return app


def main():
    # --- Python 3.14 compatibility fix ---
    # Python 3.14 stopped auto-creating an event loop in the main thread.
    # python-telegram-bot's run_polling() still expects one to exist,
    # so we create it ourselves here before anything else runs.
    try:
        asyncio.get_event_loop()
    except RuntimeError:
        asyncio.set_event_loop(asyncio.new_event_loop())

    app = build_application()
    logger.info("Bot starting... press Ctrl+C to stop.")
    app.run_polling()


if __name__ == "__main__":
    main()

File: src/bot.js
/**
 * bot.js
 * ------
 * Direct port of bot.py. Same philosophy as the original: a thin front
 * door that greets people into the Mini App, plus a handful of admin
 * text-command fallbacks. Everything else lives in the Mini App
 * (server.js + webapp/), same as before.
 *
 * One library difference worth knowing up front: python-telegram-bot
 * passes a `context` object into every handler that carries the bot
 * instance, command arguments, etc. bot.js's node-telegram-bot-api is
 * lower-level — a handler only gets `(msg, match)`, where `match` is
 * whatever your regex captured. There's no built-in "context.args"; the
 * regex itself has to capture the argument text, which is why you'll see
 * patterns like /^\/broadcast(?:\s+(.+))?$/ below instead of a plain
 * "/broadcast" command name.
 */

const fs = require("fs");
const TelegramBot = require("node-telegram-bot-api");
const repo = require("./repository");

// ---- Load bot token ----
function loadToken() {
  if (process.env.BOT_TOKEN) return process.env.BOT_TOKEN;
  if (fs.existsSync("token.txt")) return fs.readFileSync("token.txt", "utf8").trim();
  throw new Error(
    "No bot token found. Create a file named token.txt with your token inside, or set the BOT_TOKEN environment variable."
  );
}

// ---- Load admin Telegram IDs ----
function rootAdminIds() {
  const idsStr = process.env.ADMIN_IDS || "";
  return new Set(
    idsStr
      .replace(/\n/g, ",")
      .split(",")
      .map((s) => s.trim())
      .filter((s) => /^\d+$/.test(s))
      .map(Number)
  );
}

// Root admins (env var, fixed at deploy time) UNION admins added on the
// fly via /addadmin (stored in Mongo) — same two-source model as the
// Python version, so any existing admin can promote someone else
// instantly without a redeploy.
async function loadAdminIds() {
  const dbIds = await repo.getAdminIdsFromDb();
  return new Set([...rootAdminIds(), ...dbIds]);
}

function isRootAdmin(telegramId) {
  const idsStr = process.env.ADMIN_IDS || "";
  const rootIdStrings = idsStr.replace(/\n/g, ",").split(",").map((s) => s.trim());
  return rootIdStrings.includes(String(telegramId));
}

const WEBAPP_URL = process.env.WEBAPP_URL || "";

function appButtonKeyboard() {
  // node-telegram-bot-api takes the keyboard as a plain object matching
  // Telegram's own Bot API JSON shape directly — no builder class the
  // way python-telegram-bot's InlineKeyboardMarkup/InlineKeyboardButton
  // wraps it, just the raw structure Telegram's API expects.
  return { reply_markup: { inline_keyboard: [[{ text: "🚀 Open App", web_app: { url: WEBAPP_URL } }]] } };
}

// Mirrors the Python @admin_only decorator, just as a plain wrapper
// function instead of a decorator (JS doesn't have decorators for plain
// functions the way Python does) — it closes over `bot` so it can send
// the "admins only" refusal itself before ever calling the real handler.
function adminOnly(bot, handler) {
  return async (msg, match) => {
    const admins = await loadAdminIds();
    if (!admins.has(msg.from.id)) {
      await bot.sendMessage(msg.chat.id, "🔒 This command is for bot admins only.");
      return;
    }
    return handler(msg, match);
  };
}

function buildBot(app) {
  const token = loadToken();
  // Switched from { polling: true } to a webhook. Polling means this
  // process itself calls Telegram's getUpdates in a loop — and Telegram
  // only allows ONE active poller per bot token at a time. Render's
  // zero-downtime deploys start the new instance before fully stopping
  // the old one, so for a few seconds both are polling simultaneously
  // — Telegram's own answer to that is exactly the 409 "terminated by
  // other getUpdates request" error. A webhook doesn't have this
  // problem at all: Telegram pushes updates to a URL instead of two
  // processes racing to ask "any updates?" — there's no "only one
  // poller" constraint to violate in the first place.
  const bot = new TelegramBot(token);

  if (WEBAPP_URL) {
    const webhookPath = `/telegram-webhook/${token}`;
    // The token in the URL path isn't for auth (Telegram doesn't send
    // any secret back) — it's so the endpoint can't be trivially
    // guessed and spammed with fake updates by someone who doesn't
    // already know your bot token, which they'd need for anything
    // meaningful anyway. Registered on `app` directly (passed in from
    // server.js) rather than through the normal routes/ layer, since
    // this isn't a /api/* route and doesn't go through authUser.
    // No explicit express.json() here — server.js already applies it
    // globally, before buildBot() ever runs, so req.body is already
    // parsed by the time a request reaches this handler. Adding it again
    // here would try to re-read an already-consumed request stream.
    app.post(webhookPath, (req, res) => {
      // Caught locally rather than relying on server.js's global error
      // handler — this route gets registered later than that handler
      // (buildBot() runs inside main(), after the error handler was
      // already set up at module-load time), so an error here wouldn't
      // actually reach it. Simpler to just not depend on registration
      // order at all than to restructure server.js's startup sequence
      // for one route.
      try {
        bot.processUpdate(req.body);
      } catch (err) {
        console.error("Error processing Telegram webhook update:", err);
      }
      res.sendStatus(200);
    });
    bot.setWebHook(`${WEBAPP_URL}${webhookPath}`).catch((err) => {
      console.error("Failed to set Telegram webhook:", err.message);
    });
  } else {
    console.warn("WEBAPP_URL is not set — the bot has no public HTTPS URL to register a webhook against, so it won't receive any messages. Set WEBAPP_URL to fix this.");
  }

  // ---------- Basic commands ----------

  bot.onText(/^\/start$/, async (msg) => {
    if (WEBAPP_URL) {
      await bot.sendMessage(
        msg.chat.id,
        "👋 Welcome to GrowthHub!\n\nFind trusted entrepreneurs and freelancers for any job, or list your own services and get discovered.\n\nTap below to get started:",
        appButtonKeyboard()
      );
    } else {
      await bot.sendMessage(msg.chat.id, "👋 Welcome! The app isn't set up yet — ask the bot owner to set the WEBAPP_URL environment variable.");
    }
  });

  bot.onText(/^\/app$/, async (msg) => {
    if (!WEBAPP_URL) {
      await bot.sendMessage(msg.chat.id, "The app isn't set up yet — the bot owner needs to set the WEBAPP_URL environment variable.");
      return;
    }
    await bot.sendMessage(msg.chat.id, "Tap below to open the app:", appButtonKeyboard());
  });

  bot.onText(/^\/help$/, async (msg) => {
    let text = "🧭 Almost everything lives in the app now — registering, browsing entrepreneurs, editing your listing, and ratings.\n\n/app — open it\n/start — welcome message";
    const admins = await loadAdminIds();
    if (admins.has(msg.from.id)) {
      text += "\n\nYou're an admin. Send /adminhelp to see admin-only commands (also available in the app's Admin tab).";
    }
    await bot.sendMessage(msg.chat.id, text);
  });

  // ---------- Admin-only commands ----------

  bot.onText(/^\/adminhelp$/, adminOnly(bot, async (msg) => {
    await bot.sendMessage(
      msg.chat.id,
      "🔧 *Admin commands*\n\n" +
        "/stats — see totals (entrepreneurs, services, ratings)\n" +
        "/broadcast <message> — message every registered entrepreneur\n" +
        "/forceremove <name> — remove any entrepreneur's listing by name\n" +
        "/addadmin <telegram_id> — make someone an admin instantly (no redeploy)\n" +
        "/removeadmin <telegram_id> — remove a bot-added admin\n" +
        "/listadmins — see everyone with admin access",
      { parse_mode: "Markdown" }
    );
  }));

  bot.onText(/^\/stats$/, adminOnly(bot, async (msg) => {
    const s = await repo.getStats();
    await bot.sendMessage(
      msg.chat.id,
      `📊 *Bot stats*\nEntrepreneurs: ${s.entrepreneurs}\nUnique services: ${s.services}\nRatings given: ${s.ratings}`,
      { parse_mode: "Markdown" }
    );
  }));

  // (?:\s+(.+))? — a non-capturing optional group: "/broadcast" alone
  // still matches (so we can show the usage message), but if there IS
  // text after it, match[1] captures everything from there to the end
  // of the message.
  bot.onText(/^\/broadcast(?:\s+([\s\S]+))?$/, adminOnly(bot, async (msg, match) => {
    const message = match[1];
    if (!message) {
      await bot.sendMessage(msg.chat.id, "Usage: /broadcast <message>");
      return;
    }
    const telegramIds = await repo.getAllTelegramIds();
    let sent = 0, failed = 0;
    for (const telegramId of telegramIds) {
      try {
        await bot.sendMessage(telegramId, `📢 Announcement:\n\n${message}`);
        sent++;
      } catch (e) {
        // Same reasons as the Python version's comment: the user blocked
        // the bot, or never actually started a chat with it.
        console.warn(`Broadcast failed for ${telegramId}:`, e.message);
        failed++;
      }
    }
    await bot.sendMessage(msg.chat.id, `Broadcast complete. Sent: ${sent}, failed: ${failed}.`);
  }));

  bot.onText(/^\/forceremove(?:\s+([\s\S]+))?$/, adminOnly(bot, async (msg, match) => {
    const name = match[1];
    if (!name) {
      await bot.sendMessage(msg.chat.id, "Usage: /forceremove <name>\nExample: /forceremove Jane Doe");
      return;
    }
    const { success, telegramId } = await repo.forceDeleteByName(name);
    if (success) {
      await bot.sendMessage(msg.chat.id, `✅ Removed the listing matching "${name}" (user ID ${telegramId}).`);
    } else {
      await bot.sendMessage(msg.chat.id, `No entrepreneur found matching "${name}".`);
    }
  }));

  bot.onText(/^\/addadmin(?:\s+(\d+))?$/, adminOnly(bot, async (msg, match) => {
    if (!match[1]) {
      await bot.sendMessage(msg.chat.id, "Usage: /addadmin <telegram_id>\nThe person you're adding must message @userinfobot to get their numeric ID first.");
      return;
    }
    const newAdminId = Number(match[1]);
    await repo.addAdmin(newAdminId, msg.from.id);
    await bot.sendMessage(msg.chat.id, `✅ ${newAdminId} is now an admin — no redeploy needed.`);
  }));

  bot.onText(/^\/removeadmin(?:\s+(\d+))?$/, adminOnly(bot, async (msg, match) => {
    if (!match[1]) {
      await bot.sendMessage(msg.chat.id, "Usage: /removeadmin <telegram_id>");
      return;
    }
    const targetId = Number(match[1]);
    if (isRootAdmin(targetId)) {
      await bot.sendMessage(msg.chat.id, "That's a root admin (set via Render's ADMIN_IDS) — remove them there instead, not here.");
      return;
    }
    const removed = await repo.removeAdmin(targetId);
    await bot.sendMessage(msg.chat.id, removed ? "✅ Removed." : "That ID wasn't a bot-added admin.");
  }));

  bot.onText(/^\/listadmins$/, adminOnly(bot, async (msg) => {
    const rootIds = [...rootAdminIds()].sort((a, b) => a - b);
    const dbIds = [...(await repo.getAdminIdsFromDb())].sort((a, b) => a - b);
    const lines = ["👑 *Root admins* (Render ADMIN_IDS):"];
    lines.push(...(rootIds.length ? rootIds.map((i) => `• ${i}`) : ["  (none set)"]));
    lines.push("\n🛠 *Added via /addadmin:*");
    lines.push(...(dbIds.length ? dbIds.map((i) => `• ${i}`) : ["  (none)"]));
    await bot.sendMessage(msg.chat.id, lines.join("\n"), { parse_mode: "Markdown" });
  }));

  // ---------- Shared contact (phone verification) ----------
  // Fires when someone taps "Verify with Telegram" in the Mini App, which
  // triggers Telegram's native contact-share prompt. Telegram itself
  // vouches for the number here — that's what makes it trustworthy,
  // versus a number just typed into a form. bot.on("contact", ...) is
  // node-telegram-bot-api's equivalent of python-telegram-bot's
  // MessageHandler(filters.CONTACT, ...).
  bot.on("contact", async (msg) => {
    const contact = msg.contact;
    if (contact.user_id !== msg.from.id) {
      await bot.sendMessage(msg.chat.id, "Please share your own contact, not someone else's.");
      return;
    }
    await repo.setVerifiedPhone(msg.from.id, contact.phone_number);
    await bot.sendMessage(msg.chat.id, "✅ Phone verified! Go back to the app — it should unlock automatically within a few seconds.");
  });

  return bot;
}

module.exports = { buildBot, loadToken, loadAdminIds, rootAdminIds, isRootAdmin, WEBAPP_URL };
